src/mlpack/methods/ann/layer/log_softmax_impl.hpp

```diff
@@ -19,7 +19,7 @@ namespace mlpack {
 namespace ann /** Artificial Neural Network. */ {
 
 template<typename InputDataType, typename OutputDataType>
-LogSoftMax<InputDataType, OutputDataType>::LogSoftMax()
+LogSoftMax<InputDataType, OutputDataType>::LogSoftMax() : deterministic(0)
```

Comment:
I don't think you need the use of `deterministic`. The `input` to `Backward` layer is the output obtained at the end of the `Forward` pass.

Reply:
I also thought that but when I saw the Jacobian Test function, 
https://github.com/mlpack/mlpack/blob/fded9ee333bfbac179e9826bdfe4a3b093ef67f2/src/mlpack/tests/ann_test_tools.hpp#L84-L96
...then I thought of doing this.

Reply:
I'm not sure what the `JacobianTest` is doing, maybe @zoq can help with this.

Reply:
Jacobian Test computes the derivative by using the first principle,
 ```f'(x) = lim (f(x + h) - f(x - h)) / 2h```and then comparing it with backward function. It is really confusing how other tests passed based on this. 😤

Reply:
Probably because the threshold is to high.

Reply:
The threshold for functions where this is tested is around `1e-3` I think. Not sure it qualifies as high.